pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/UselessStringValueOfRule.java

```diff
@@ -45,6 +47,13 @@ public Object visit(ASTPrimaryPrefix node, Object data) {
                                 && ((ASTReferenceType) argType.getChild(0)).isArray()) {
                             return super.visit(node, data);
                         }
+                        // declaration may be of format [TYPE] [IDENTIFIER] [LBRACKET] [RBRACKET]
+                        try {
+                            GenericToken t = ((ASTPrimitiveType)argType.getChild(0)).jjtGetFirstToken();
+                            if (t.getNext().getNext().getKind() == 79 && t.getNext().getNext().getNext().getKind() == 80) {
```

Comment:
That's very hard to understand... (and it would fail, if the code would be `String.valueOf(otherString.toCharArray(), 0, 10)`)
I think, the problem is, that the rule doesn't check the number or arguments of the method call "String.valueOf".
In line 30/32 we figured, that this is a String.valueOf call. In line 37/39, we figured, that there are arguments - and exactly here we can check the number of arguments (`args.size()`) - the rule should IMHO only consider calls to String.valueOf with exactly one argument - which then automatically ignores "String.valueOf(char[], int, int)".

Reply:
I think the issue more stems from how the PMD Java parser works. The problem is that the parser sees `char buf[]` as a primitive `char` type, not an array, since the identifier comes before the brackets. You can write a test case with just `char buf[]` as the parameter to the function and it would fail in that case. I tried playing around with the parser to get it to properly recognize the brackets but couldn't get it to work.